File: 백준/Silver/1654. 랜선 자르기/랜선 자르기.py
import sys
input = sys.stdin.readline
k, n = map(int, input().split())
l = []
res = 0

for _ in range(k):
    l.append(int(input()))
    
e = max(l)
s = 1

while s <= e:
    mid = (s + e) // 2
    m = 0
    for i in l:
        m += i // mid
    if m >= n:
        res = mid
        s = mid + 1
    else:
        e = mid-1
        
print(res)

File: 백준/Silver/1654. 랜선 자르기/랜선 자르기.java
import java.util.*;
import java.io.*;

public class Main {
    public static void main(String[] args) throws Exception {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st = new StringTokenizer(br.readLine(), " ");
        int n = Integer.parseInt(st.nextToken());
        int k = Integer.parseInt(st.nextToken());
        long s = 1;
        long e = 0;
        int[] arr = new int[n];
        for (int i = 0; i < n; i++) {
            arr[i] = Integer.parseInt(br.readLine());
            if (e < arr[i]) {
                e = arr[i];
            }
        }
        long res = -1;
        while (s <= e) {
            long mid = (s + e) / 2;
            int total = 0;
            for (int i = 0; i < n; i++) {
                total += arr[i] / mid;
            }
            if (total >= k) {
                s = mid + 1;
                res = Math.max(res, mid);
            } else {
                e = mid -1;

            }
        }
        System.out.println(res);
    }
}
